airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py

```diff
@@ -0,0 +1,257 @@
+        1 to expected retries attempts.
+        """
+        if self._backoff_strategy.max_retries is not None:
+            max_tries = max(0, self._backoff_strategy.max_retries) + 1
+
+        user_backoff_handler = user_defined_backoff_handler(max_tries=max_tries, max_time=self._backoff_strategy.max_time)(self._send)
+        backoff_handler = default_backoff_handler(
+            max_tries=max_tries, max_time=self._backoff_strategy.max_time, factor=self._backoff_strategy.retry_factor
+        )
+        # backoff handlers wrap _send, so it will always return a response
+        response = backoff_handler(user_backoff_handler)(request, request_kwargs)
+
+        return response
+
+    def _send(self, request: requests.PreparedRequest, request_kwargs: Optional[Mapping[str, Any]] = None) -> requests.Response:
+
+        self._logger.debug(
+            "Making outbound API request", extra={"headers": request.headers, "url": request.url, "request_body": request.body}
+        )
+
+        if request_kwargs is None:
+            request_kwargs = {}
+
+        try:
+            response: requests.Response = self._session.send(request, **request_kwargs)
+            self._response_decoder.validate_response(response)
+        except requests.RequestException as exc:
+            response_action = ResponseAction.RETRY
+            failure_type = FailureType.transient_error
+            response_error_message = f"Request to {request.url} failed with exception: {exc}"
+            self._logger.debug(response_error_message)
+            exc = exc
+            response = None
+        else:
+            response_action, failure_type, response_error_message = self._error_handler.interpret_response(response=response)
+            exc = None
+
+        # Evaluation of response.text can be heavy, for example, if streaming a large response
+        # Do it only in debug mode
+        if response is not None:
+            if self._logger.isEnabledFor(logging.DEBUG):
+                self._logger.debug(
+                    "Receiving response", extra={"headers": response.headers, "status": response.status_code, "body": response.text}
+                )
+
+            if response_action == ResponseAction.FAIL:
+                error_message = (
+                    f"Request failed to {request.url} with error {exc}"
+                    if exc
+                    else response_error_message
+                    or f"Request to {response.request.url} failed with status code {response.status_code} and error message {self._error_message_parser.parse_response_error_message(response)}"
+                )
+                # TODO: Provide better internal/external error messaging
+                raise AirbyteTracedException(
+                    internal_message=error_message,
+                    message=error_message,
+                    failure_type=failure_type,
+                )
+
+            if response_action == ResponseAction.IGNORE:
```

Comment:
Trying to see this as "there is an error on the HTTP request level but I want to take action elsewhere", it still wasn't the best as I would end up have to parse again the response in order to understand the error. In the end, I think it made more sense [to raise domain level exceptions](https://github.com/airbytehq/airbyte/pull/37853/files#diff-bdfb79faf4a4584afa7a26bc905115a5d481f9cb3d91ddde0f885007807d4c01R94) that can be managed [at the domain level](https://github.com/airbytehq/airbyte/pull/37853/files#diff-0572edeedd561901f8110c0f88b48efde87d1ed91247c0eb189c7229da12090dR386) without having to bother about HTTP. 

I'm wondering what is the vision for this. Is there another way to use the IGNORE action?

Reply:
errors are sometimes ignored when we know they can happen but there's nothing we can do about them.

An example use case would is when a substream returns 404 when there is no data to fetch for a given parent record.

Reply:
Re: 404 for substream. I'm wondering how this would be handled by a connector --- for example if we had substreams where we knew 404s were possible but want to ignore as well as "standard" streams where we want to fail on 404s -- this would require a custom implementation of the error handler for the substream cases, corrwect?

Reply:
yes, different streams would use a different mapping of error code -> action